### PR TITLE
[AIR/train] Use new Train API

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Uses the new AIR Train API for examples and tests.

The `Result` object gets a new attribute - `log_dir`, pointing to the Trial's `logdir` allowing users to access tensorboard logs and artifacts of other loggers.

This PR only deals with "low hanging fruit" - tests that need substantial rewriting or Train user guide are not touched. Those will be updated in followup PRs.

Tests and examples that concern deprecated features or which are duplicated in AIR have been removed or disabled.

Requires https://github.com/ray-project/ray/pull/25943 to be merged in first

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by Yard1** [comment]
One example fail will be fixed by https://github.com/ray-project/ray/pull/25849

File: python/ray/train/tests/test_examples.py [hunk 1]
```diff
@@ -19,50 +20,49 @@
...
-    result = results[0]
+    result = results.metrics
 
-    loss = result["loss"]
-    assert len(loss) == epochs
-    assert loss[-1] < loss[0]
-
-    accuracy = result["accuracy"]
-    assert len(accuracy) == epochs
-    assert accuracy[-1] > accuracy[0]
```
Comment (amogkam) [inline python/ray/train/tests/test_examples.py:53]: Can we keep these assert statements? We should still test that accuracy improves right?

File: python/ray/train/tests/test_examples.py [hunk 2]
```diff
@@ -88,24 +88,24 @@ def test_tensorflow_mnist_fail(ray_start_2_cpus):
...
+    trainer = TorchTrainer(
+        linear_train_func,
+        train_loop_config=config,
+        scaling_config=dict(num_workers=num_workers),
+    )
+    results = trainer.fit()
 
-    for result in results:
-        assert len(result) == epochs
-        assert result[-1]["loss"] < result[0]["loss"]
```
Comment (amogkam) [inline python/ray/train/tests/test_examples.py:105]: Same here- can we keep these assert statements?

File: python/ray/train/tests/test_minimal.py [hunk 3]
```diff
@@ -53,27 +43,23 @@ def train_func():
...
-    assert len(test_callback.result_list) == value
-    assert len(test_callback.result_list[0]) == num_workers
-    print(test_callback.result_list[0])
-    assert all(result[key] == value for result in test_callback.result_list[0])
-
-    # Test checkpointing.
-    assert trainer.latest_checkpoint[key] == value
-
-    trainer.shutdown()
+    assert results.checkpoint
```
Comment (amogkam) [inline python/ray/train/tests/test_minimal.py:62]: should be 
```suggestion
    assert results.checkpoint == checkpoint
```
?

**Review by amogkam — CHANGES_REQUESTED** [review]
Thanks a lot for all the hard work @Yard1!

Left some comments, mostly making sure we are not losing any test coverage with this PR.

File: python/ray/train/tests/test_examples.py (hunk 1 above)
Comment (Yard1) [reply python/ray/train/tests/test_examples.py:53]: We only have access to the last result in the `Result` object :/

File: python/ray/train/tests/test_tune.py [hunk 4]
```diff
@@ -96,18 +115,7 @@ def test_tune_tensorflow_mnist(ray_start_8_cpus):
...
 
-def test_tune_error(ray_start_2_cpus):
-    def train_func(config):
-        raise RuntimeError("Error in training function!")
-
-    trainer = Trainer(TestConfig(), num_workers=1)
-    TestTrainable = trainer.to_tune_trainable(train_func)
-
-    with pytest.raises(TuneError):
-        tune.run(TestTrainable)
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/train/tests/test_tune.py:107]: This should be tested in Tuner tests

[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/train/tests/test_tune.py:107]: Just checked. This test will fail, because we have `raise_on_failed_trial=False` inside `TunerInternal`. Is this correct? cc @xwjiang2010 @krfricke

File: python/ray/train/tests/test_minimal.py (hunk 3 above)
Comment (Yard1) [reply python/ray/train/tests/test_minimal.py:62]: checkpoints cannot be compared directly, and the current checkpoint id will differ. I'll test for equality of `key value` though

File: python/ray/train/tests/test_callbacks.py [hunk 5]
```diff
@@ -1,357 +0,0 @@
-from typing import Dict, List
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/train/tests/test_callbacks.py:1]: Hmm, got it

File: python/ray/train/tests/test_examples.py (hunk 1 above)
Comment (Yard1) [reply python/ray/train/tests/test_examples.py:53]: I am not sure what's the best approach here. I do agree we should test for this. I can think of either creating a debug trainer that will return the trial dataframe, or we should revamp Trainer (and possibly Tune) so that the entire Trial history is returned in a `Result`.

File: python/ray/util/ml_utils/checkpoint_manager.py [hunk 6]
```diff
@@ -132,7 +132,7 @@ def to_air_checkpoint(self) -> Optional[Checkpoint]:
             checkpoint_dir = TrainableUtil.find_checkpoint_dir(checkpoint_data)
             checkpoint = Checkpoint.from_directory(checkpoint_dir)
         elif isinstance(checkpoint_data, bytes):
-            with tempfile.mkdtemp() as tmpdir:
+            with tempfile.TemporaryDirectory() as tmpdir:
```
Comment (Yard1) [inline python/ray/util/ml_utils/checkpoint_manager.py:135]: cc @krfricke

Comment (krfricke) [reply python/ray/util/ml_utils/checkpoint_manager.py:135]: Resolved on master, thanks!

File: python/ray/train/tests/test_examples.py (hunk 1 above)
Comment (Yard1) [reply python/ray/train/tests/test_examples.py:53]: Changes in https://github.com/ray-project/ray/pull/25943 will allow to test this again

File: python/ray/train/tests/test_tune.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/train/tests/test_tune.py:107]: The error raised by tune.run will be caught by the Tuner, if `raise_on_failed_trial=True`

[the remark this answers is not in the document]
Comment (xwjiang2010) [reply python/ray/train/tests/test_tune.py:107]: ah I see, we are supposed to check `Result` object for the actually error message I think.

[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/train/tests/test_tune.py:107]: It's not raising any exception right now (seems to be only logging/printing it?). That doesn't seem right

[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/train/tests/test_tune.py:107]: Should this be done in the Trainer then?

[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/train/tests/test_tune.py:107]: Ok got it

File: python/ray/train/tests/test_examples.py (hunk 1 above)
Comment (Yard1) [reply python/ray/train/tests/test_examples.py:53]: added back in

File: python/ray/train/tests/test_examples.py (hunk 2 above)
Comment (Yard1) [reply python/ray/train/tests/test_examples.py:105]: added back in

**Review by amogkam — COMMENTED** [review]
Thanks @Yard1!

But I'm thinking- for the existing Ray Train tests, should we keep these tests until we fully remove the old APIs? Instead of changing them, just duplicating them for AIR?

**Comment by Yard1** [comment]
@amogkam I am not sure what is our policy on deprecated APIs being tested. By including the tests in CI we implicate that we are at the very least maintaining old APIs. If we are going to hard deprecate them soon, I am not sure if it's worth extra runtime. But on the other hand, adding tests back in is easy and I am happy to do that (with the examples being left out).

File: python/ray/train/train_loop_utils.py [hunk 7]
```diff
@@ -178,18 +180,29 @@ def load_checkpoint() -> Optional[Dict]:
...
             for iter in range(checkpoint["epoch"], 5):
                 print(iter)
 
-        trainer = Trainer(backend="torch")
-        trainer.start()
-        trainer.run(train_func, checkpoint={"epoch": 3})
+        checkpoint = Checkpoint.from_dict(
+            {
+                # this would be set during checkpoint saving
+                "_current_checkpoint_id": 1,
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/train/train_loop_utils.py:194]: Yeah, the decision on that had changed since I first edited them. I will revert

File: python/ray/tune/result_grid.py [hunk 8]
```diff
@@ -180,6 +180,7 @@ def _trial_to_result(self, trial: Trial) -> Result:
             checkpoint=checkpoint,
             metrics=trial.last_result.copy(),
             error=self._populate_exception(trial),
+            log_dir=trial.local_dir,
```
Comment (matthewdeng) [inline python/ray/tune/result_grid.py:183]: Should this be `trial.logdir` instead?

File: python/ray/train/examples/torch_fashion_mnist_example.py [hunk 9]
```diff
@@ -2,14 +2,14 @@
 from typing import Dict
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/train/examples/torch_fashion_mnist_example.py:2]: Yeah we should dedupe them in a future PR!

File: python/ray/tune/result_grid.py (hunk 8 above)
Comment (Yard1) [reply python/ray/tune/result_grid.py:183]: So the intention was to point to a place where tensorboard would be accesible from (so the `local_dir` of the trial). On the other hand, that information will also be contained in `trial.logdir` and accessible by getting its parent. I guess we could go with that to avoid confusion

File: python/ray/air/result.py [hunk 10]
```diff
@@ -37,6 +38,7 @@ class Result:
     metrics: Optional[Dict[str, Any]]
     checkpoint: Optional[Checkpoint]
     error: Optional[Exception]
+    log_dir: Optional[str]
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/air/result.py:41]: done

File: python/ray/tune/result_grid.py (hunk 8 above)
Comment (Yard1) [reply python/ray/tune/result_grid.py]: I have switched it to using `trial.logdir`

File: python/ray/train/BUILD [hunk 11]
```diff
@@ -73,25 +64,15 @@ py_test(
...
-    main = "examples/tune_cifar_pytorch_pbt_example.py",
-    srcs = ["examples/tune_cifar_pytorch_pbt_example.py"],
+    main = "examples/tune_cifar_torch_pbt_example.py",
+    srcs = ["examples/tune_cifar_torch_pbt_example.py"],
     tags = ["team:ml", "exclusive", "pytorch", "tune"],
     deps = [":train_lib"],
     args = ["--smoke-test"]
 )
 
-py_test(
```
Comment (amogkam) [inline python/ray/train/BUILD:85]: Do we need to run the renamed example in CI?

File: python/ray/train/examples/horovod/horovod_example.py [hunk 12]
```diff
@@ -142,21 +144,21 @@ def train_func(config):
...
-    loss_per_epoch = trainer.run(train_func, config=kwargs)
-    trainer.shutdown()
-    print(loss_per_epoch)
+    trainer = HorovodTrainer(
+        train_func,
+        train_loop_config=kwargs,
+        scaling_config={"use_gpu": use_gpu, "num_workers": num_workers},
+    )
+    results = trainer.fit()
+    print(results)
```
Comment (amogkam) [inline python/ray/train/examples/horovod/horovod_example.py:161]: Nit: only print the metrics ant not the entire result?

File: python/ray/train/examples/mlflow_fashion_mnist_example.py [hunk 13]
```diff
@@ -1,20 +1,23 @@
...
+        train_loop_config={"lr": 1e-3, "batch_size": 64, "epochs": 4},
+        scaling_config={"num_workers": num_workers, "use_gpu": use_gpu},
+        run_config=RunConfig(
+            callbacks=[MLflowLoggerCallback(experiment_name="train_fashion_mnist")]
+        ),
     )
+    final_results = trainer.fit()
 
-    print("Full losses for rank 0 worker: ", final_results)
+    print("Full results for rank 0 worker: ", final_results)
```
Comment (amogkam) [inline python/ray/train/examples/mlflow_fashion_mnist_example.py:20]: Nit: only print the metrics and not the entire result?

File: python/ray/train/examples/tensorflow_linear_dataset_example.py [hunk 14]
```diff
@@ -71,22 +70,20 @@ def train_func(config):
...
+        train_func,
+        train_loop_config={"lr": 1e-3, "batch_size": 32, "epochs": 4},
+        datasets=datasets,
+        dataset_config=dataset_configs,
+        scaling_config={"num_workers": num_workers, "use_gpu": use_gpu},
     )
-    trainer.shutdown()
-    print(f"Results: {results[0]}")
+    results = trainer.fit()
+    print(f"Results: {results}")
```
Comment (amogkam) [inline python/ray/train/examples/tensorflow_linear_dataset_example.py:86]: Nit: only print the metrics and not the entire result?

Comment (amogkam) [reply python/ray/train/examples/tensorflow_linear_dataset_example.py:86]: Same thing for all the examples

File: python/ray/train/examples/torch_fashion_mnist_example.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/train/examples/torch_fashion_mnist_example.py:2]: +1 we should make sure to dedup and move over the AIR examples to train in a follow up

**Review by amogkam — APPROVED** [review]
Thanks @Yard1 lgtm overall!

Please ping again when this is ready to merge!

File: python/ray/train/BUILD (hunk 11 above)
Comment (Yard1) [reply python/ray/train/BUILD:85]: It is being ran in AIR CI already as `tune_torch_linear_dataset_example`

**Comment by Yard1** [comment]
@amogkam it's ready to merge!
